### ISSUE TITLE
Custom MigrationResolver to provide files for SQL migrations should be practicable

### ISSUE THREAD

**Issue by ben-peacock-uk** [flyway/flyway#3393]
###### Which version and edition of Flyway are you using?
8.4.4 Teams Edition (trial)
 
 
###### If this is not the latest version, can you reproduce the issue with the latest one as well? (Many bugs are fixed in newer releases and upgrading will often resolve the issue)
n/a
 
 
###### Which client are you using? (Command-line, Java API, Maven plugin, Gradle plugin)
Maven plugin
 
 
###### Which database are you using? (Type & version)
Oracle 19
 
 
###### Which operating system are you using?
MacOS
 
 
###### What did you do? (Please include the content causing the issue, any relevant configuration settings, the SQL statement(s) that failed (if any), and the command you ran)
 
Tried to write a MigrationResolver to convert system files to Flyway migrations as follows:
- infer that file is repeatable from subdirectory rather than file name prefix
- provide unique description for files with different suffixes without requiring artificial file name
- order files based on subdirectory and suffix without requiring artificial names
- (if practicable - order based on dependencies to other files)
 
e.g. from this file structure
+ packages
  - test.pkb
  - test.pks
+ views
  - test.sql
  - test1.sql
 
Convert to repeatable migrations in this order with these descriptions (or similar unique descriptions)
- test view (views/test.sql)
- test1 view (views/test1.sql)
- test spec (packages/test.pks)
- test body (packages/test.pkb)
 
https://github.com/flyway/flyway/issues/1608 appears to provide example of such a MigrationResolver but relies on internals that have now changed significantly. The comment closing that issue says it can be done via configuration, but that only applies to being able to read sql files with other suffixes. File names still require a prefix and to be unique (excluding suffix) across all subdirectories, and order is based on file name across all subdirectories.
 
https://github.com/flyway/flyway/issues/1388 raises a similar requirement and the solution suggested was to write a custom MigrationResolver, but not clear how.
 
###### What did you expect to see?
 
Able to write custom MigrationResolver and/or set configuration to infer migration details from file structure rather than require artificial file name prefixes and suffixes.
 
e.g. there could be an executor factory api to allow an SqlMigrationExecutor to be created from a Resource and Configuration. Or executor could be optional and default provided based on migration details.
 
###### What did you see instead?
 
Unable to write such a custom MigrationResolver. Need to provide custom Executor which seems impractical. The existing SqlMigrationExecutor would be fine but has package-private constructor. I can copy it but it requires SqlScriptExecutorFactory and SqlScript from internal classes, which require their own set of internal objects to create.
 
Workaround was to set properties:
- repeatableSqlMigrationPrefix=
- sqlMigrationSeparator=_
- sqlMigrationSuffixes=.sql,.pkb,.pks
 
And rename files to:
+ packages
   - _test.pks
   - _test_body.pkb
+ views
   - _0_test.sql
   - _0_test1.sql
 
This workaround may seem fairly trivial. But we have many hundreds of such files in such structures across several schemas, which we are looking to bring under Flyway control. And these file names seem very messy. And updating views with new dependencies may require order change, which requires file rename, which then requires repair to the schema history to delete the file with the old name.

**Comment by DoodleBobBuffPants** [comment]
In V9, a `MigrationResolver` will also have access to `Configuration`, `ResourceProvider`, `SqlScriptFactory` and `SqlScriptExecutorFactory` - would this help solve this?

**Comment by ben-peacock-uk** [comment]
Yes thanks, that would help. I will have to try it when I can get hold of v9. I think I will still have to make a copy of SqlMigrationExecutor, and rely on some internal classes, but it might become practicable.
